### PR TITLE
Update getting started guides

### PR DESCRIPTION
**Pull request by jcapriot**
Update and detail processes for working with SimPEG

This updates many of our "getting started" guides to recommend good practices.

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -49,31 +52,75 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
+This being said, all simpeg developers and admins are essentially volunteers
+providing there time for the benefit of the community. This does mean that
+it might take some time for us to get your PR.
+
 
 Licensing
-*********
+=============
 
 The contributed code will be licensed under SimPEG's license
```
Comment (prisae) [inline CONTRIBUTING.rst:125]: I would rephrase this to: «[...] will be licensed under the MIT license [...]»

The term «SimPEG's license» sounds too much like it is its own license. And what we should avoid at any cost in open-source is own-baked / custom licenses. Naming it by its name avoids confusion.

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -49,31 +52,75 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
+- 100% (or as close as possible) difference testing. Meaning any new code is completely tested.
+- All tests are passing.
+- All reviewer comments (if any) have been addressed.
+- A developer approves the PR.
+
+After all these steps are satisfied, a ``@simpeg/simpeg-admin`` will merge your pull request into
+the main branch (feel free to ping one of us on github).
+
+This being said, all simpeg developers and admins are essentially volunteers
+providing there time for the benefit of the community. This does mean that
```
Comment (prisae) [inline CONTRIBUTING.rst:118]: "there" -> "their"

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -49,31 +52,75 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
+it might take some time for us to get your PR.
+
 
 Licensing
-*********
+=============
```
Comment (prisae) [inline CONTRIBUTING.rst:123]: underline 4 characters too long (ok, very picky here :-) )

File: README.rst [hunk 4]
```diff
@@ -103,30 +103,49 @@ If you are using the electromagnetics module of SimPEG, please cite:
         doi = "http://dx.doi.org/10.1016/j.cageo.2017.06.018"
     }
 
+Questions
+=========
+
+Please do not create an issue to ask a question. You will get a faster
+response by posting on our Discourse discussion forum:
+https://simpeg.discourse.group/. If you prefer real-time chat,
+you can join our slack group at http://slack.simpeg.xyz.
```
Comment (prisae) [inline README.rst:113]: Maybe phrase it positively (first the way to do it, and at the end what NOT to do)? 

«The fastest way go get a response to your answers is by posting on our Discourse discussion forum: https://simpeg.discourse.group/. If you prefer real-time chat,
you can join our slack group at http://slack.simpeg.xyz. Please do not create an issue to ask a question.»

File: docs/content/basic/installing.rst [hunk 5]
```diff
@@ -28,12 +28,14 @@ Installing SimPEG
...
 
-You can install SimPEG using the `conda package manager <https://conda.io/>`_ that comes with the Anaconda distribution:
+You can install SimPEG using the `conda package manager <https://conda.io/>`_ that
+comes with the Anaconda distribution:
 
 .. code::
 
     conda install SimPEG --channel conda-forge
 
+Installing through `conda` is our recommended method of installation.
```
Comment (prisae) [inline docs/content/basic/installing.rst:39]: People seem to get frustrated with the speed of conda, for good reason. By teaching and generally I started to recommend mamba by default. And in particular `mambaforge`, as you then have your environment already setup to always use `--channel conda-forge`.
https://github.com/conda-forge/miniforge#mambaforge

I don't know if that is something that could be mentioned here.

File: docs/content/basic/installing_for_developers.rst [hunk 6]
```diff
@@ -78,40 +80,29 @@ repository, have a look at the article: `A successful git branching model
...
-Open a terminal in this directory and create a symlink for SimPEG ::
-
-    ln -s ../SimPEG/SimPEG .
-
-Then, in your shell, you need to add a :code:`PYTHONPATH` variable. For Mac and
-Linux, if you are using Z shell (`Oh My Zsh <http://ohmyz.sh/>`_ is used by a
-lot of SimPEG developers) or bash open the config in a text editor, ie::
-
-    nano ~/.zshrc
+To get started developing SimPEG we recommend setting up an environment with Anaconda_
```
Comment (prisae) [inline docs/content/basic/installing_for_developers.rst:83]: Again, I think we should probably not push on "anaconda" - in the end, our codes are not even available on anaconda, they are available on conda-forge. We should strengthen conda-forge (mambaforge).

File: docs/content/basic/practices.rst [hunk 7]
```diff
@@ -8,34 +8,30 @@ Practices
 
 Here we cover
 
-- :ref:`Testing <testing>`
-- :ref:`Style <style>`
-- :ref:`Licensing <licensing>`
+- `testing`_
+- style_
+- licensing_
```
Comment (prisae) [inline docs/content/basic/practices.rst:13]: Are there intentionally backticks by testing, but not by style nor by licensing?

**Review by prisae — APPROVED** [review]
Awesome developments, I like all these recent additions! I just skimmed briefly through it and left a few comments. Some are rather personal opinion and can easily be ignored if you do not like them ;-)

File: README.rst (hunk 4 above)
Comment (jcapriot) [reply README.rst:113]: I like it.

File: docs/content/basic/installing.rst (hunk 5 above)
Comment (jcapriot) [reply docs/content/basic/installing.rst:39]: I think it is worthwhile dropping a link to the `mamba` project here in a ``.. note::` section.
Also, I don't think `mamba` defaults to having the `-c conda-forge` option always on anymore. I've noticed this on a few of the more recent installs I've done using it. It also makes sense considering they are billing it as a "drop-in" replacement for `conda`.

Comment (prisae) [reply docs/content/basic/installing.rst:40]: No, `mamba` does not have it by default I think. But if you install Python using `mambaforge`, it makes the necessary changes so that you do have the `conda-forge` channel.

Comment (jcapriot) [reply docs/content/basic/installing.rst:40]: Ah yes.

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -49,34 +56,81 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
 
 Licensing
-*********
+=========
 
-The contributed code will be licensed under SimPEG's license
-(https://github.com/simpeg/simpeg/blob/main/LICENSE). If you did not write
+The code contributed to SimPEG will be licensed under the `MIT license
+<https://github.com/simpeg/simpeg/blob/main/LICENSE>`_ which allows open
+and commercial use and extension of SimPEG. If you did not write
```
Comment (santisoler) [inline CONTRIBUTING.rst:133]: I think it would be nice to make it clearer **when** the code contributed to SimPEG becomes licensed under the MIT License. The _will_ might lead to think that this code will _eventually_ be MIT licensed, but without giving a clear time in the future.

In my opinion, once any contributor opens a PR, that code is already being shared under the MIT license. It's the author that licenses their contribution under that license, and not the package maintainers by merging their code.

```suggestion
All code contributed to SimPEG is licensed under the `MIT license
<https://github.com/simpeg/simpeg/blob/main/LICENSE>`_ which allows open
and commercial use and extension of SimPEG. If you did not write
```

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -49,34 +56,81 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
+
+.. _testing: https://docs.simpeg.xyz/content/basic/practices.html#testing
+
+.. _documentation: https://docs.simpeg.xyz/content/basic/practices.html#documentation
+
+.. _code style: https://docs.simpeg.xyz/content/basic/practices.html#style
+
 If you have an idea for how to improve SimPEG, please first create an issue
-and :ref:`suggest an enhancement <suggesting_enhancements>`. We will use the
+and `suggesting_enhancements`_. We will use the
```
Comment (santisoler) [inline CONTRIBUTING.rst:82]: Minor edit:
```suggestion
and `suggest enhancements <#suggesting-enhancements>`__. We will use the
```

File: docs/content/basic/installing_for_developers.rst [hunk 10]
```diff
@@ -38,23 +37,26 @@ Working with Git and GitHub
 
 
 To keep track of your code changes and contribute back to SimPEG, you will
-need a github_ account and fork the `SimPEG repository <http://github.com/simpeg/simpeg>`_
+need a github_ account, then fork the `SimPEG repository <http://github.com/simpeg/simpeg>`_
+to your local account.
 (`How to fork a repo <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_). Software
```
Comment (santisoler) [inline docs/content/basic/installing_for_developers.rst:42]: Remove "Software" from this sentence.
```suggestion
(`How to fork a repo <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_).
```

File: docs/content/basic/installing_for_developers.rst [hunk 11]
```diff
@@ -78,40 +80,39 @@ repository, have a look at the article: `A successful git branching model
...
-
-    echo $PYTHONPATH
-
-the output should be::
-
-    /Users/USER/git/python_symlinks
+    pip uninstall SimPEG
 
+We no longer recommend modifying your python path environment variable as a way
+to install SimPEG for developers.
```
Comment (santisoler) [inline docs/content/basic/installing_for_developers.rst:115]:
```suggestion

.. note::

    We no longer recommend modifying your python path environment variable as a way
    to install SimPEG for developers.
```

File: docs/content/basic/practices.rst [hunk 12]
```diff
@@ -8,34 +8,30 @@ Practices
...
-    :width: 80px
-
 On each update, SimPEG is tested using the continuous integration service
-`Travis CI <https://travis-ci.org/>`_. We use `Codecov <http://codecov.io>`_
-to check and provide stats on how much of the code base is covered by tests.
-This tells which lines of code have been run in the test suite. It does not
-tell you about the quality of the tests run! In order to assess that, have a
-look at the tests we are running - they tell you the assumptions that we do
-not want to break within the code base.
+`azure pipelines <https://azure.microsoft.com/en-us/products/devops/pipelines>`_.
```
Comment (santisoler) [inline docs/content/basic/practices.rst:29]:
```suggestion
`Azure Pipelines <https://azure.microsoft.com/en-us/products/devops/pipelines>`_.
```

File: docs/content/basic/practices.rst [hunk 13]
```diff
@@ -177,7 +162,9 @@ Documentation
 -------------
 
 Documentation helps others use your code! Please document new contributions.
-SimPEG uses `sphinx <http://www.sphinx-doc.org/>`_ to build the documentation.
+SimPEG trys to follow the `numpydoc` style of docstrings, check out the
```
Comment (santisoler) [inline docs/content/basic/practices.rst:165]:
```suggestion
SimPEG tries to follow the `numpydoc` style of docstrings, check out the
```

File: docs/content/basic/practices.rst [hunk 14]
```diff
@@ -220,11 +228,29 @@ For example:
...
+We also actively update the code base to ensure pep8 compliance by checking with
+`flake8 <https://flake8.pycqa.org/>`_ This performs style checks that could lead
+towards bugs, performs checks on consistent documentation formatting, or just
+identify poor coding practices. This is an ongoing process where we are fixing one
+style warning at a time. The fixed style warnings are checked to ensure no new code
+goes against an already established style. This test can also be installed locally
+using pre-commit hooks, similar to `black` above.
+
+
+If you have sublime text 3, the linter can be set up through sublime text 3's package manager. The step by
```
Comment (santisoler) [inline docs/content/basic/practices.rst:251]: I would put all these tips related to specific text editors inside their own subsection. They are related to the code style, but they are not part of the style rules that SimPEG follows. Replace Sublime Text with any other editor and the style check that SimPEG follows are the same.


As an alternative, what if we put all the content related to text editors in a separate page? Instead of including tips on how to configure them within the pages where the style rules and testing standards are laid out, move them to their own page. I think it has two benefits:

* It would be easier to update the text editor pages if some configuration gets outdated.
* It separates the "this is the standards we use to develop high quality code" and the "this is how you should configure your text editor". I think contributors might want to read these two topics in different moments (in one they learn, the other one is a how-to guide).

**Review by santisoler — CHANGES_REQUESTED** [review]
Nice job with this @jcapriot! It's so refreshing not seeing the instructions for manually adding a directory to PYTHONPATH.

I left a few comments, most of them some minor corrections. Let me know what you think!

File: docs/content/basic/installing_for_developers.rst [hunk 15]
```diff
@@ -38,23 +37,26 @@ Working with Git and GitHub
...
+to your local account.
 (`How to fork a repo <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_). Software
 
 
 .. _github: http://github.com
 
-Next, clone your fork so that you have a local copy. We recommend setting up a
-directory called :code:`git` in your home directory to put your version-
-controlled repositories. There are two ways you can clone a repository: (1)
+Next, clone your fork to your computer so that you have a local copy. We recommend setting up a
```
Comment (santisoler) [inline docs/content/basic/installing_for_developers.rst:47]: I think it would look better if we use some RST to split the terminal and the desktop client instructions. Something like this:
```rst
.. _github: http://github.com

.. _SourceTree: https://www.sourcetreeapp.com/

.. _GitKraken: https://www.gitkraken.com/

Next, clone your fork to your computer so that you have a local copy. We recommend setting up a
directory in your home directory to put your version-controlled repositories (e.g.  called :code:`git`).
There are two ways you can clone a repository:

1. From a terminal (checkout: https://docs.github.com/en/get-started/quickstart/set-up-git for an tutorial):

   .. code-block:: bash

      git clone https://github.com/YOUR-USERNAME/SimPEG

2. Using a desktop client such as SourceTree_ or GitKraken_.

   .. image:: ../../images/sourceTreeSimPEG.png
      :align: center
      :width: 400
      :target: https://www.sourcetreeapp.com/

   If this is your first time managing a github_ repository through SourceTree_,
   it is also handy to set up the remote account so it remembers your github_
   user name and password

   .. image:: ../../images/sourceTreeRemote.png
      :align: center
      :width: 400

   For managing your copy of SimPEG and contributing back to the main
   repository, have a look at the article: `A successful git branching model
   <http://nvie.com/posts/a-successful-git-branching-model/>`_

```

File: docs/content/basic/practices.rst (hunk 14 above)
Comment (jcapriot) [reply docs/content/basic/practices.rst:251]: At the very least they should be separated. However, I might be in favor of just removing these details all together. They are too specific for these areas to really be helpful, and why should we highlight one text editor over another?

Comment (jcapriot) [reply docs/content/basic/practices.rst:251]: I also do not personally use sublime, so I actually have no clue if any of these recommendations are even relevant.

Comment (santisoler) [reply docs/content/basic/practices.rst:251]: I agree with your points. I don't think we should recommend one text editor over the rest. And personally, I wouldn't recommend one that is not open-source as the first option.

I think that we should remove the text editor sections, unless someone has a strong reason to keep them and is willing to keep them up-to-date.

File: CONTRIBUTING.rst (hunk 8 above)
Comment (jcapriot) [reply CONTRIBUTING.rst]: That's a good point, the language is quite subtle in general here.

File: docs/content/basic/installing_for_developers.rst (hunk 10 above)
Comment (jcapriot) [reply docs/content/basic/installing_for_developers.rst]: I wish black could format `rst` files for line-length.

File: CONTRIBUTING.rst [hunk 16]
```diff
@@ -49,34 +45,92 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
+.. _pr_stages:
+
+Stages of a pull request
+------------------------
+
+When first creating a pull request, try to make your suggested changes as tightly
+scoped as possible. The fewer changes you make, the faster your branch will be
+merged!
+
+If your pull request is not ready for final review, but you still want feedback
```
Comment (lheagy) [inline CONTRIBUTING.rst:106]:
```suggestion
If your pull request (PR) is not ready for final review, but you still want feedback
```

Comment (jcapriot) [reply CONTRIBUTING.rst:106]: I add the abbreviation a little higher up.

File: CONTRIBUTING.rst [hunk 17]
```diff
@@ -49,34 +45,92 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
 enhancement, we certainly appreciate if you are also willing to take action
 and start a pull request!
 
+.. _questions:
+
+Questions
+---------
```
Comment (lheagy) [inline CONTRIBUTING.rst:51]: should this be at the same level as "Issues" (rather than under it)? Since we ask people not to create issues for questions, it might be simpler as its own heading

File: CONTRIBUTING.rst [hunk 18]
```diff
@@ -49,34 +45,92 @@ remember that SimPEG is maintained on a volunteer basis. If you suggest an
...
+
+If you have a question regarding a specific use of SimPEG, the fastest way
+to get a response is by posting on our Discourse discussion forum:
+https://simpeg.discourse.group/. Alternatively, if you prefer real-time chat,
+you can join our slack group at http://slack.simpeg.xyz.
+Please do not create an issue to ask a question.
+
```
Comment (lheagy) [inline CONTRIBUTING.rst:58]: Also, should we include a heading on meetings? This is a great way for folks who are interested in contributing to engage with us.

File: SimPEG/regularization/base.py [hunk 19]
```diff
@@ -634,6 +634,53 @@ def _multiplier_pair(self):
...
 
 class WeightedLeastSquares(ComboObjectiveFunction):
+    r"""Weighted least squares measure on model smallness and smoothness.
+
+    L2 regularization with both smallness and smoothness (first order
+    derivative) contributions.
+
+    Parameters
+    ----------
+    mesh : discretize.base.BaseMesh
```
Comment (lheagy) [inline SimPEG/regularization/base.py:644]:
```suggestion
    mesh : discretize.base.BaseMesh
        The mesh on which the model parameters are defined. This is used for constructing difference operators for the smoothness terms. 
```

Comment (lheagy) [reply SimPEG/regularization/base.py:644]: just a suggestion, but might be helpful

File: docs/content/basic/practices.rst [hunk 20]
```diff
@@ -44,59 +40,48 @@ but meant to provide a few places to look when you are developing and would
...
-            vertMap = maps.SurjectVertical1D(M)
-            combo = expMap*vertMap
-            m = np.arange(3.0)
-            t_true = np.exp(np.r_[0,0,1,1,2,2.])
-            self.assertLess(np.linalg.norm((combo * m)-t_true,np.inf),TOL)
+    import numpy as np
+    import discretize
+    from SimPEG import maps
+
+    def test_mapMultiplication(self):
```
Comment (lheagy) [inline docs/content/basic/practices.rst:77]: do we want to update this to use more pep8-compliant naming? `test_map_multiplication` 

and M --> `mesh` etc?

**Review by lheagy — APPROVED** [review]
This looks great! Thanks for going through and bringing these pages up to date. I left a few minor comments in there

File: CONTRIBUTING.rst (hunk 17 above)
Comment (jcapriot) [reply CONTRIBUTING.rst:51]: That makes sense.

File: CONTRIBUTING.rst (hunk 18 above)
Comment (jcapriot) [reply CONTRIBUTING.rst:58]: I think I'll add it in README.txt so it's a little more front and center.

File: docs/content/basic/practices.rst (hunk 20 above)
Comment (jcapriot) [reply docs/content/basic/practices.rst:77]: might as well!

File: SimPEG/regularization/base.py (hunk 19 above)
Comment (jcapriot) [reply SimPEG/regularization/base.py:644]: Yeah, more detail is better. The only reason I added it here was because it was already being used as an example of documentation in the style guide.
